File: sdk/python/hookline/ws.py
"""HookLine asyncio WebSocket client with auto-reconnect."""

from __future__ import annotations

import asyncio
import json
import logging
from collections.abc import Callable, Awaitable
from typing import Any

logger = logging.getLogger(__name__)

_RECONNECT_BASE = 1.0
_RECONNECT_MAX = 60.0
_RECONNECT_FACTOR = 2.0


class HookLineWS:
    """Asyncio WebSocket client for HookLine real-time events.

    Usage::

        async with HookLineWS("ws://localhost:8080", "my-api-key") as ws:
            await ws.subscribe("orders.*")
            ws.on_event(handle_event)
            await asyncio.sleep(3600)

    Or manage the lifecycle manually::

        ws = HookLineWS("ws://localhost:8080", "my-api-key")
        await ws.connect()
        try:
            await ws.subscribe("orders.*")
            ...
        finally:
            await ws.close()
    """

    def __init__(self, base_url: str, api_key: str) -> None:
        base = base_url.rstrip("/").replace("http://", "ws://").replace("https://", "wss://")
        self._ws_url = f"{base}/v1/ws?token={api_key}"
        self._api_key = api_key
        self._ws: Any = None
        self._running = False
        self._reconnect_delay = _RECONNECT_BASE
        self._event_callbacks: list[Callable[[dict], Awaitable[None] | None]] = []
        self._presence_callbacks: list[Callable[[dict], Awaitable[None] | None]] = []
        self._read_task: asyncio.Task | None = None

    # ── Context manager ───────────────────────────────────────────────────

    async def __aenter__(self) -> "HookLineWS":
        await self.connect()
        return self

    async def __aexit__(self, *_: Any) -> None:
        await self.close()

    # ── Public API ────────────────────────────────────────────────────────

    async def connect(self) -> None:
        """Open the WebSocket connection and start the read loop."""
        self._running = True
        await self._do_connect()
        self._read_task = asyncio.create_task(self._read_loop())

    async def close(self) -> None:
        """Close the WebSocket connection and stop the read loop."""
        self._running = False
        if self._read_task is not None:
            self._read_task.cancel()
            try:
                await self._read_task
            except asyncio.CancelledError:
                pass
        if self._ws is not None:
            await self._ws.close()
            self._ws = None

    async def subscribe(self, topic: str) -> None:
        """Subscribe to a topic pattern (e.g. ``orders.*``)."""
        await self._send({"type": "subscribe", "topic": topic})

    async def unsubscribe(self, topic: str) -> None:
        """Unsubscribe from a topic pattern."""
        await self._send({"type": "unsubscribe", "topic": topic})

    async def publish(self, topic: str, payload: dict | None = None) -> None:
        """Publish an event over the WebSocket connection."""
        await self._send({"type": "publish", "topic": topic, "payload": payload or {}})

    async def mark_read(self, event_id: str) -> None:
        """Acknowledge receipt of an event."""
        await self._send({"type": "mark_read", "event_id": event_id})

    def on_event(self, callback: Callable[[dict], Awaitable[None] | None]) -> None:
        """Register a callback invoked for each incoming event frame."""
        self._event_callbacks.append(callback)

    def on_presence(self, callback: Callable[[dict], Awaitable[None] | None]) -> None:
        """Register a callback invoked for each presence update frame."""
        self._presence_callbacks.append(callback)

    # ── Internals ─────────────────────────────────────────────────────────

    async def _do_connect(self) -> None:
        try:
            import websockets  # type: ignore[import]
        except ImportError as exc:
            raise ImportError(
                "The 'websockets' package is required for HookLineWS. "
                "Install it with: pip install websockets"
            ) from exc

        self._ws = await websockets.connect(self._ws_url)
        self._reconnect_delay = _RECONNECT_BASE
        logger.debug("HookLineWS connected to %s", self._ws_url)

    async def _read_loop(self) -> None:
        while self._running:
            try:
                assert self._ws is not None
                raw = await self._ws.recv()
                await self._dispatch(json.loads(raw))
            except asyncio.CancelledError:
                return
            except Exception as exc:
                if not self._running:
                    return
                logger.warning("HookLineWS error: %s — reconnecting in %.1fs", exc, self._reconnect_delay)
                await asyncio.sleep(self._reconnect_delay)
                self._reconnect_delay = min(self._reconnect_delay * _RECONNECT_FACTOR, _RECONNECT_MAX)
                try:
                    await self._do_connect()
                except Exception as conn_exc:
                    logger.error("HookLineWS reconnect failed: %s", conn_exc)

    async def _dispatch(self, frame: dict) -> None:
        ftype = frame.get("type")
        if ftype == "event":
            for cb in self._event_callbacks:
                result = cb(frame)
                if asyncio.iscoroutine(result):
                    await result
        elif ftype == "presence":
            for cb in self._presence_callbacks:
                result = cb(frame)
                if asyncio.iscoroutine(result):
                    await result

    async def _send(self, msg: dict) -> None:
        if self._ws is None:
            raise RuntimeError("HookLineWS is not connected")
        await self._ws.send(json.dumps(msg))

File: sdk/go/hookline/ws.go
package hookline

import (
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"sync"
	"sync/atomic"
	"time"

	"github.com/gorilla/websocket"
)

// WSEvent is an event message received over the WebSocket.
type WSEvent struct {
	Type      string                 `json:"type"`
	ID        string                 `json:"id"`
	Topic     string                 `json:"topic"`
	Payload   map[string]interface{} `json:"payload"`
	Timestamp int64                  `json:"timestamp"`
}

// PresenceUpdate is a presence notification received over the WebSocket.
type PresenceUpdate struct {
	Type   string   `json:"type"`
	Event  string   `json:"event"`
	UserID string   `json:"user_id"`
	Topics []string `json:"topics,omitempty"`
	Ts     int64    `json:"ts"`
}

// AckMessage is an acknowledgement for a published message.
type AckMessage struct {
	Type     string `json:"type"`
	ClientID string `json:"client_id,omitempty"`
	EventID  string `json:"event_id"`
}

// WSOptions configures a WSClient.
type WSOptions struct {
	Topics     []string
	UserID     string
	OnEvent    func(WSEvent)
	OnPresence func(PresenceUpdate)
	OnAck      func(AckMessage)
	OnReconnect func()
	OnError    func(error)
}

// WSClient is a goroutine-safe WebSocket client with auto-reconnect.
type WSClient struct {
	url  string
	opts WSOptions

	mu   sync.Mutex
	conn *websocket.Conn

	closed         int32 // atomic; 1 = permanently disconnected
	reconnectDelay time.Duration
}

// newWSClient creates a new WSClient. Use Client.Connect instead of calling
// this directly.
func newWSClient(baseURL, apiKey string, opts WSOptions) *WSClient {
	wsURL := strings.Replace(baseURL, "http", "ws", 1)
	wsURL = fmt.Sprintf("%s/v1/ws?token=%s", wsURL, apiKey)
	if len(opts.Topics) > 0 {
		wsURL += "&topics=" + strings.Join(opts.Topics, ",")
	}
	return &WSClient{
		url:            wsURL,
		opts:           opts,
		reconnectDelay: time.Second,
	}
}

// Connect dials the server and starts the read loop in a goroutine.
func (c *WSClient) Connect() error {
	conn, _, err := websocket.DefaultDialer.Dial(c.url, nil)
	if err != nil {
		return err
	}
	c.mu.Lock()
	c.conn = conn
	c.mu.Unlock()
	go c.readLoop()
	return nil
}

// Publish sends a publish message to the server.
func (c *WSClient) Publish(topic string, payload interface{}, clientID string) error {
	return c.send(map[string]interface{}{
		"type":      "publish",
		"topic":     topic,
		"payload":   payload,
		"client_id": clientID,
	})
}

// Subscribe adds topics to the server-side filter.
func (c *WSClient) Subscribe(topics []string) error {
	return c.send(map[string]interface{}{
		"type":   "subscribe",
		"topics": topics,
	})
}

// Unsubscribe removes topics from the server-side filter.
func (c *WSClient) Unsubscribe(topics []string) error {
	return c.send(map[string]interface{}{
		"type":   "unsubscribe",
		"topics": topics,
	})
}

// MarkRead sends a read receipt for an event.
func (c *WSClient) MarkRead(eventID string) error {
	return c.send(map[string]interface{}{
		"type":     "read",
		"event_id": eventID,
	})
}

// Disconnect permanently closes the connection (no auto-reconnect).
func (c *WSClient) Disconnect() {
	atomic.StoreInt32(&c.closed, 1)
	c.mu.Lock()
	if c.conn != nil {
		_ = c.conn.Close()
	}
	c.mu.Unlock()
}

// send serialises msg to JSON and writes it as a text frame.
func (c *WSClient) send(msg interface{}) error {
	data, err := json.Marshal(msg)
	if err != nil {
		return err
	}
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.conn == nil {
		return errors.New("hookline: not connected")
	}
	return c.conn.WriteMessage(websocket.TextMessage, data)
}

func (c *WSClient) readLoop() {
	for {
		c.mu.Lock()
		conn := c.conn
		c.mu.Unlock()

		_, data, err := conn.ReadMessage()
		if err != nil {
			if atomic.LoadInt32(&c.closed) == 1 {
				return
			}
			if c.opts.OnError != nil {
				c.opts.OnError(err)
			}
			c.scheduleReconnect()
			return
		}
		c.dispatch(data)
	}
}

func (c *WSClient) dispatch(data []byte) {
	var raw map[string]json.RawMessage
	if err := json.Unmarshal(data, &raw); err != nil {
		return
	}
	var msgType string
	if t, ok := raw["type"]; ok {
		_ = json.Unmarshal(t, &msgType)
	}
	switch msgType {
	case "event":
		if c.opts.OnEvent != nil {
			var evt WSEvent
			if json.Unmarshal(data, &evt) == nil {
				c.opts.OnEvent(evt)
			}
		}
	case "presence":
		if c.opts.OnPresence != nil {
			var p PresenceUpdate
			if json.Unmarshal(data, &p) == nil {
				c.opts.OnPresence(p)
			}
		}
	case "ack":
		if c.opts.OnAck != nil {
			var ack AckMessage
			if json.Unmarshal(data, &ack) == nil {
				c.opts.OnAck(ack)
			}
		}
	}
}

func (c *WSClient) scheduleReconnect() {
	delay := c.reconnectDelay
	time.AfterFunc(delay, func() {
		if atomic.LoadInt32(&c.closed) == 1 {
			return
		}
		conn, _, err := websocket.DefaultDialer.Dial(c.url, nil)
		if err != nil {
			if c.opts.OnError != nil {
				c.opts.OnError(err)
			}
			// double the delay and retry again
			c.reconnectDelay = min(c.reconnectDelay*2, 30*time.Second)
			c.scheduleReconnect()
			return
		}
		c.mu.Lock()
		c.conn = conn
		c.mu.Unlock()
		c.reconnectDelay = time.Second // reset on success
		if c.opts.OnReconnect != nil {
			c.opts.OnReconnect()
		}
		go c.readLoop()
	})
	c.reconnectDelay = min(delay*2, 30*time.Second)
}

func min(a, b time.Duration) time.Duration {
	if a < b {
		return a
	}
	return b
}
